### PR TITLE
List View: Allow dragging to all levels of the block hierarchy

### LINKED ISSUE #33678: List view drag and drop - make it possible to drop at all levels of the hierarchy
## Description
At the moment, there's a known issue when dragging and dropping in List View. It's only possible to drag a block where there's an adjacent block at the position you wish to drop to.

For example, in the following structure, a paragraph can't be moved to after the group because the drag and drop system uses the presence of other blocks to determine the drop target:
Screenshot 2021-07-26 at 1 46 11 pm

Technically this is difficult to achieve because List View is currently implemented as a flat table. 

## Step-by-step reproduction instructions
1. Create a nested block structure with no blocks after the parent block
2. Try dragging and dropping one of the nested blocks after the parent

## Expected behaviour
It should be possible to drop a block after the parent

## Actual behaviour
This isn't possible if the parent block is the last block in a list, only when there's another a block after the parent block

## Video

https://user-images.githubusercontent.com/1813435/155583396-d20bd450-f9a7-4b91-ad15-7c66067ee327.mp4

### PR DESCRIPTION
**Pull request by andrewserong**
## What?

Fixes #33678

This PR seeks to add the ability to drag to all levels of the block hierarchy within the list view. This is most notable when attempting to drag beneath an expanded Group block, particularly if it is heavily nested.

## Why?

It's important that folks can drag beneath a nested block. If we can allow dragging to each level of the block hierarchy, then it should allow for a more powerful drag and drop interface. Enabling this behaviour is a prerequisite for many of the current ideas for visual improvements to drag and drop — first we need to make sure it's possible to accurately determine the drop target, before we can improve how it looks.

## How?

* Calculate the desired nesting level based on the mouse cursor position, and using a hard-coded value to account for the indent for each nesting level (if anyone has a good idea about how to measure this instead, I'm happy to explore it!)
* Construct a list of parents for the target block, based on the block data and matching based on `rootClientId`s
* Update nesting logic so that it also uses a calculation based on the current nesting level to determine where the mouse cursor will be recognised as a nesting gesture
* Factor in dragged blocks when determining the nesting level of the next block to compare
* Determine the `blockIndex` for the dropped block based on either being at the very end of the parent (if the subsequent block is not at the same level) or by using the `blockIndex` of the next block (if the subsequent block is at the same level)
* Add some extra tests to try to cover most of the changes with unit tests

## To-do

* [x] ~Ensure the `blockIndex` is the correct one for where the user is dragging. That is, it should be the last index of the parent block _or_ one less that the subsequent block if the drag position has the same level as the subsequent block.~
* [x] ~Ensure dragging to the root level position still works correctly~
* [x] ~If the block being dragged is the last of the parent's children, then currently it interferes with the check of the last visual child in the list view (as technically it isn't the last child).~
* [x] ~Fix existing tests, add additional tests~
* [ ] Fine-tune the positioning of where each nesting level begins (currently it's hard-coded, but can we determine it based on anything within the DOM?)
* [ ] Ensure this doesn't break if a list view is using a custom `rootClientId` — I haven't tested this yet, but I _think_ this should work okay

## Testing Instructions

<details>

<summary>Some heavily nested block markup</summary>

```html
<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph -->
<p>A paragraph in a group level 1</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph -->
<p>A paragraph in a group level 2</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph -->
<p>A paragraph in a group level 3</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph -->
<p>A paragraph in a group level 4</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph -->
<p>A paragraph in a group level 5</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"anchor":""} -->
<p>A paragraph in a group level 6</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"anchor":""} -->
<p>A paragraph in a group level 7</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"anchor":""} -->
<p>A paragraph in a group level 8</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->

<!-- wp:heading -->
<h2 class="wp-block-heading">A heading</h2>
<!-- /wp:heading -->

<!-- wp:paragraph -->
<p>A paragraph</p>
<!-- /wp:paragraph -->
```
</details>

1. Add a series of nested Group blocks to a post.
2. Try dragging a block beneath the bottom nested block across all levels of the hierarchy (you need to be hovered over the bottom half of the nested block)
3. Add additional blocks at each level of the hierarchy
4. Attempt a lot of different kinds of dragging and see if the updated logic breaks anything

### CODE REVIEW

**Comment by andrewserong** [comment]
The code's not perfect for this one (it relies on a hard-coded value for the nesting level), but I think this PR is probably at a good place for review / feedback on the approach.

The overall goal with this PR is to see if we can improve the experience for dragging through all levels of the block hierarchy. My hope is that this PR should improve things a fair bit on its own, but also unlock some of the recent discussions and ideas for how to improve drag and drop in the list view visually — the step taken in this PR is to more accurately determine which level of the block hierarchy to drag to.

Happy for any feedback or ideas! In terms of scope — this PR is just about solving dragging through the different levels of the block hierarchy. I'm hoping we can look at visual changes separately (i.e. here's an exploration into changing the drag chip: https://github.com/WordPress/gutenberg/pull/49820).

**Comment by jasmussen** [comment]
Nice, just from a quick smell test, this feels substantially better and less fiddly:

dragon drop

* Separate, but nice dropzone indicator matching the in-canvas one.
* Scrollbars working, and coming into their own.
* Doesn't seem to suffer from that _click and move ever so slightly_ to invoke a drag and drop, which trunk has.

Seems like a solid step forward. Nice!

**Comment by noisysocks** [comment]
I couldn't get this to work at all when testing in Firefox unfortunately:

https://user-images.githubusercontent.com/612155/235383388-606513c3-256f-47b3-84fe-396fa3daba82.mp4

It works exactly as advertised when testing in Chrome though, so you must be running into some browser discrepancy. I didn't check Edge—probably a good idea to run this through Browserstack before merging.

Definitely less finicky than before. I like it.

More of a question for our design team and shouldn't block this PR at all but I wonder if we should display gutter lines or alternative the background colour of each leaf in the list view so that you can trace a line up with your eye to see what level you're dragging into.

Will read through the code now.

File: packages/block-editor/src/components/list-view/use-list-view-drop-zone.js
Comment (noisysocks) [inline packages/block-editor/src/components/list-view/use-list-view-drop-zone.js:63]: I suppose we could use [`getComputedStyle()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle) to get the actual padding used. It might be slow though? Definitely benchmark it if you go down that route.

I don't really hate hardcoding this. To me, the level of indentation is a key part of what `ListView` communicates to the user and therefore is arguably "content". I don't think it would be inappropriate at all to set `marign-left` on the expander using an inline style, rewrite [this logic](https://github.com/WordPress/gutenberg/blob/36e3b335bf08c8d321d32a444ae1e5e50eef56e3/packages/block-editor/src/components/list-view/style.scss#L379-L388) using JavaScript, and then that way the same code that's used for setting the `margin-left` can be used here for calculating the drop destination.

Practically, all that we _really_ care about, is preventing a future developer from updating the CSS I linked to above without also updating this constant. So at the very minimum, put a comment above the CSS that refers to this constant and vice versa.

**Comment by andrewserong** [comment]
> I couldn't get this to work at all when testing in Firefox unfortunately:

Oh, weird! It's not working in Firefox for me, either, but it is working in Safari, Edge, and Chrome. I'll dig in. Thanks for catching that!

File: packages/block-editor/src/components/list-view/use-list-view-drop-zone.js
Comment (noisysocks) [inline packages/block-editor/src/components/list-view/use-list-view-drop-zone.js:101]: You have nice doc comments on all the functions except this one. Makes me feel bad for it 🥹

File: packages/block-editor/src/components/list-view/use-list-view-drop-zone.js
Comment (noisysocks) [inline packages/block-editor/src/components/list-view/use-list-view-drop-zone.js:123]: Is it worth updating `getListViewDropTarget` and all its helper functions to have `blocksData` be a `Map` (keyed by `clientId`) instead of an array so that we don't need lookups like this `find()`? I'm not too familiar with this code but it looks like `getCandidateBlockParents` will be called when the user moves the mouse so I imagine performance could be a problem if there are ~1000s of blocks.

Only bother if it makes an actual difference 😀 (premature optimisation yada yada)

Comment (talldan) [reply packages/block-editor/src/components/list-view/use-list-view-drop-zone.js:123]: When I originally wrote it, I'm sure `blockData` was supposed to only be computed once when the user starts dragging. I see it's now computed on every dragover event. 🤔 

Anyway, my suggestion was going to be to store the parent client ids in `blocksData` ahead of time, but I think the only real optimization would come from ensuring that `blocksData` isn't computed so often.

Comment (talldan) [reply packages/block-editor/src/components/list-view/use-list-view-drop-zone.js:123]: I tried a fix for this in this PR - https://github.com/WordPress/gutenberg/pull/50210

Comment (andrewserong) [reply packages/block-editor/src/components/list-view/use-list-view-drop-zone.js:123]: Thanks Dan! I like that direction — just left a comment on that PR about whether we then need to add data attributes to the placeholder rows that are used as part of the windowing logic. I think that should still be pretty doable, while preserving the performance improvement?

File: packages/block-editor/src/components/list-view/use-list-view-drop-zone.js
Comment (andrewserong) [reply packages/block-editor/src/components/list-view/use-list-view-drop-zone.js:101]: I was determined to make sure these were all documented, so thanks for flagging that one, it did look lonely! Updated in https://github.com/WordPress/gutenberg/pull/49742/commits/ca8ec6d55b11ebf05df39e1ac662415cdce9209a

File: packages/block-editor/src/components/list-view/use-list-view-drop-zone.js
Comment (andrewserong) [reply packages/block-editor/src/components/list-view/use-list-view-drop-zone.js:63]: Good idea to add comments to link these together (added in https://github.com/WordPress/gutenberg/pull/49742/commits/5e9068ce86d244dd23a9ca966b57d6c1c72fb7b0). I also really like the idea of switching the indentation over to JS styles so that the hard-coded indentation level is used in both places 👍

If / when this PR lands, I'm happy to dig into that as a follow-up!

**Comment by andrewserong** [comment]
Thanks again for reviewing @noisysocks — the Firefox bug took me a little longer than anticipated to debug, but I got there in the end. TIL it turns out that accessing `ariaLevel` on an element via the Element API isn't supported in Firefox — https://caniuse.com/mdn-api_element_arialevel! I've swapped that usage out for including a `data-level` attribute that gets used in a similar way to how we're passing down the `expanded` state, and it appears to be working well for me so far 👍

I also like the idea for the performance improvement of keying `blocksData` by `clientId`. I probably lean _slightly_ toward looking into that as a code quality follow-up PR, but also happy to look into that now if folks think it's worth ensuring this change is performant enough.

**Review by noisysocks — APPROVED** [review]
Lovely! Nice work.

**Comment by noisysocks** [comment]
> I also like the idea for the performance improvement of keying `blocksData` by `clientId`. I probably lean _slightly_ toward looking into that as a code quality follow-up PR, but also happy to look into that now if folks think it's worth ensuring this change is performant enough.

It looks like in the other thread @talldan decided that the performance impact isn't significant.

**Comment by talldan** [comment]
> TIL it turns out that accessing ariaLevel on an element via the Element API isn't supported in Firefox 

This is so random! How about `element.getAttribute( 'aria-level' )`? From what I can tell that works, and it might be preferable to adding a new attribute.

**Comment by andrewserong** [comment]
> This is so random! How about element.getAttribute( 'aria-level' )? From what I can tell that works, and it might be preferable to adding a new attribute.

Good call — I think this is preferable to relying on extra data attributes. I've swapped out the data attribute for `.getAttribute( 'aria-level' )` and re-tested in FF, Chrome, Safari, and Edge, and it's all still working correctly for me.

Were there any other changes you wanted to see made before this lands @talldan? I'm happy to dig into a few code quality improvements in follow-ups (in particular I really like Rob's idea of moving to JS for the left margins so that the logic there is consolidated).

File: packages/block-editor/src/components/list-view/use-list-view-drop-zone.js
Comment (talldan) [reply packages/block-editor/src/components/list-view/use-list-view-drop-zone.js:63]: > I don't think it would be inappropriate at all to set marign-left on the expander using an inline style, rewrite this logic using JavaScript, and then that way the same code that's used for setting the margin-left can be used here for calculating the drop destination.

I wrote it in a similar way originally but using spacer divs (which would possibly have made this PR easier), it was changed here - https://github.com/WordPress/gutenberg/pull/32063. 

Another option could be injecting css vars for the level and indent, and then doing a `calc` to multiply them.

`list-view/index.js` also has `BLOCK_LIST_ITEM_HEIGHT` that could be improved in a similar way.

Sounds good for a follow-up.

**Comment by talldan** [comment]
@andrewserong This looks good to me, thanks so much for working on this!

File: packages/block-editor/src/components/list-view/use-list-view-drop-zone.js
Comment (andrewserong) [reply packages/block-editor/src/components/list-view/use-list-view-drop-zone.js:63]: Good ideas, I'll dig into this stuff in follow-ups. Thanks!

**Comment by andrewserong** [comment]
Thanks for the reviews and feedback! Much appreciated 🙇

**Comment by scruffian** [comment]
I wonder if this would help at all with https://github.com/WordPress/gutenberg/issues/46994

**Comment by andrewserong** [comment]
> I wonder if this would help at all with https://github.com/WordPress/gutenberg/issues/46994

I don't think so, this PR is really just looking at how we determine the correct block drop target from within the list view based on the mouse cursor position relative to the list view items. I think #46994 would likely need to be solved for from the editor canvas drop zone side of things. Dave had a good idea over in https://github.com/WordPress/gutenberg/issues/46994#issuecomment-1408839130 to potential add context / metadata to the drag operation, and check this from within the editor canvas, which could be a good way to go about it. I've added a comment on that issue.

**Pull request merged**
